internal/server/pool_test.go

```diff
@@ -53,7 +54,455 @@ func TestPoolSpike(t *testing.T) {
+		cfg.MaxOpen = 3
+		cfg.MaxIdle = 3
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn2, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 3, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+		p.ReleaseConn(conn2)
+
+		be.Equal(t, 3, p.Stats().OpenConnections)
+		be.Equal(t, 3, p.Stats().Idle)
+
+		p.SetMaxIdleConns(1)
+
+		be.Equal(t, 1, p.Stats().OpenConnections)
+		be.Equal(t, 1, p.Stats().Idle)
+	})
+
+	t.Run("maxIdle > maxOpen", func(t *testing.T) {
+		cfg := genBasePoolConfig()
+		cfg.MaxOpen = 2
+		cfg.MaxIdle = 2
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+		be.Equal(t, 2, p.Stats().Idle)
+
+		p.SetMaxIdleConns(5)
+
+		be.Equal(t, 2, p.maxIdle)
+	})
+}
+
+func TestMaxIdleTime(t *testing.T) {
+	cfg := genBasePoolConfig()
+	cfg.MaxOpen = 3
+	cfg.MaxIdle = 3
+	cfg.MaxIdleTime = 2 * time.Second
+
+	rt := be.Relaxed(t)
+
+	p, err := NewPool(cfg)
+	be.NilErr(rt, err)
+	defer func() { be.NilErr(rt, p.Close()) }()
+
+	conn0, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	conn1, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+
+	p.ReleaseConn(conn0)
+	p.ReleaseConn(conn1)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	time.Sleep(cfg.MaxIdleTime + time.Second)
+
+	be.Equal(t, 0, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+	be.Equal(t, 2, p.Stats().MaxIdleTimeClosed)
+}
+
+func TestMaxIdleTimeUpdate(t *testing.T) {
+	cfg := genBasePoolConfig()
+	cfg.MaxOpen = 3
+	cfg.MaxIdle = 3
+	cfg.MaxIdleTime = 2 * time.Second
+
+	rt := be.Relaxed(t)
+
+	p, err := NewPool(cfg)
+	be.NilErr(rt, err)
+	defer func() { be.NilErr(rt, p.Close()) }()
+
+	conn0, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	conn1, err := p.AcquireConn()
+	be.NilErr(rt, err)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+
+	p.ReleaseConn(conn0)
+	p.ReleaseConn(conn1)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	time.Sleep(100 * time.Millisecond)
+
+	be.Equal(t, 2, p.Stats().OpenConnections)
+	be.Equal(t, 2, p.Stats().Idle)
+
+	p.SetConnMaxIdleTime(100 * time.Millisecond)
+
+	time.Sleep(p.maxIdleTime + time.Second)
+
+	be.Equal(t, 0, p.Stats().OpenConnections)
+	be.Equal(t, 0, p.Stats().Idle)
+	be.Equal(t, 2, p.Stats().MaxIdleTimeClosed)
+}
```

Comment:
Nit: I'd change a couple of things here in this test to make sure we're not killing the connections due to the original MaxIdleTime:
1. Set the original `MaxIdleTime` to something much much bigger. Let's say a minute?
2. Measure the time right before line 247 and at the very end to make sure that we have not consumed the whole original `MaxIdleTime`.

Reply:
Good point